components/brave_wallet/browser/bitcoin/bitcoin_wallet_service.cc

```diff
@@ -0,0 +1,647 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/bitcoin/bitcoin_wallet_service.h"
+
+#include <stdint.h>
+#include <deque>
+#include <set>
+
+#include "base/check.h"
+#include "base/functional/bind.h"
+#include "base/memory/scoped_refptr.h"
+#include "base/notreached.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/sys_byteorder.h"
+#include "base/types/expected.h"
+#include "brave/components/brave_wallet/browser/bitcoin/bitcoin_database_synchronizer.h"
+#include "brave/components/brave_wallet/browser/bitcoin/bitcoin_transaction_database.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/keyring_service.h"
+#include "brave/components/brave_wallet/common/bitcoin_utils.h"
+#include "brave/components/brave_wallet/common/hash_utils.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+constexpr uint32_t kSigHashAll = 1;
+
+std::vector<std::string> BitcoinKeyringsForNetwork(
+    const std::string& network_id) {
+  DCHECK(brave_wallet::IsBitcoinNetwork(network_id));
+  if (network_id == brave_wallet::mojom::kBitcoinMainnet) {
+    return {brave_wallet::mojom::kBitcoinKeyring84Id};
+  } else if (network_id == brave_wallet::mojom::kBitcoinTestnet) {
+    return {brave_wallet::mojom::kBitcoinKeyring84TestId};
+  } else {
+    NOTREACHED();
+  }
+
+  return {};
+}
+
+// TODO(apaymyshev): implement Push* as 'serialize to network' class
+void Push8AsLE(uint8_t i, std::vector<uint8_t>& to) {
+  to.push_back(i);
+}
+
+void Push16AsLE(uint16_t i, std::vector<uint8_t>& to) {
+  i = base::ByteSwapToLE16(i);
+  base::span<uint8_t> data_to_insert(reinterpret_cast<uint8_t*>(&i), sizeof(i));
+  to.insert(to.end(), data_to_insert.begin(), data_to_insert.end());
+}
+
+void Push32AsLE(uint32_t i, std::vector<uint8_t>& to) {
+  i = base::ByteSwapToLE32(i);
+  base::span<uint8_t> data_to_insert(reinterpret_cast<uint8_t*>(&i), sizeof(i));
+  to.insert(to.end(), data_to_insert.begin(), data_to_insert.end());
+}
+
+void Push64AsLE(uint64_t i, std::vector<uint8_t>& to) {
+  i = base::ByteSwapToLE64(i);
+  base::span<uint8_t> data_to_insert(reinterpret_cast<uint8_t*>(&i), sizeof(i));
+  to.insert(to.end(), data_to_insert.begin(), data_to_insert.end());
+}
+
+void PushVectorAsLE(const std::vector<uint8_t>& v, std::vector<uint8_t>& to) {
+  to.insert(to.end(), v.rbegin(), v.rend());
+}
+
+void PushVarInt(uint64_t i, std::vector<uint8_t>& to) {
+  if (i < 0xfd) {
```

Comment:
Please add [CompactSize Unsigned Integers](https://developer.bitcoin.org/reference/transactions.html#compactsize-unsigned-integers) to comment

Reply:
https://github.com/brave/brave-core/pull/17978/commits/c697f80928661571502650868b76f2f6252f298b